### PR TITLE
Extract HiveClientPool as an abstract class to facilitate testing

### PR DESCRIPTION
**Pull request by dbw9580**
### What changes are proposed in this pull request?

Extract `HiveClientPool` as an abstract class to facilitate testing.

### Why are the changes needed?

Currently there is no way to inject a mocked `IMetaStoreClient` into `HiveDatabase`. Adding the ability to do so facilitates unit testing.

### Does this PR introduce any user facing changes?

No.

### CODE REVIEW

**Comment by HelloHorizon** [comment]
@dbw9580 what is the test you want to add with hive metastore mock? @apc999 I remembered you mentioned we have some best practice guidelines about the mock tests, do you mind taking a look?

**Comment by dbw9580** [comment]
> what is the test you want to add with hive metastore mock

I'm adding unit tests for the group mount PR #13654. I need to mock a hive metastore client to emulate a hive database and its tables, so that I can test the group mount behaviour is working as expected.

**Comment by dbw9580** [comment]
> I need to mock a hive metastore client to emulate a hive database and its tables

The exact thing I want to do is write a `TestHiveClientPool` that extends the new `HiveClientPool` abstract class, and creates and returns mocked `IMetaStoreClient` from its `acquireClientResource`. With the mocked client, I can provide arbitrary tables and partitions to `HiveDatabase`, and I can test if the group mount behaviour proposed in #13654 is working correctly.

I have investigated some alternatives to the approach proposed in this PR. One is to use the compatibility layer provided by `HMSShim` and `CompatibleMetastoreClient`. They provide a way similar to a mocked client that can be used to return test databases and tables. But they themselves lack unit tests, and it is less straight forward if they are used in other components' unit tests

File: table/server/underdb/hive/src/main/java/alluxio/table/under/hive/HiveDatabase.java [hunk 1]
```diff
@@ -85,6 +86,16 @@ private HiveDatabase(UdbContext udbContext, UdbConfiguration configuration,
     mClientPool = CLIENT_POOL_CACHE.getPool(connectionUri);
```
Comment (apc999) [inline table/server/underdb/hive/src/main/java/alluxio/table/under/hive/HiveDatabase.java:86]:
```java
private HiveDatabase(UdbContext udbContext, UdbConfiguration configuration,
      String connectionUri, String hiveDbName)  {
  this(udbContext, configuration, connectionUri, hiveDbName, CLIENT_POOL_CACHE.getPool(connectionUri));
}
```

File: table/server/underdb/hive/src/main/java/alluxio/table/under/hive/HiveDatabase.java [hunk 2]
```diff
@@ -85,6 +86,16 @@ private HiveDatabase(UdbContext udbContext, UdbConfiguration configuration,
...
   }
 
+  @VisibleForTesting
+  HiveDatabase(UdbContext udbContext, UdbConfiguration configuration,
+               HiveClientPool clientPool, String connectionUri, String hiveDbName) {
```
Comment (apc999) [inline table/server/underdb/hive/src/main/java/alluxio/table/under/hive/HiveDatabase.java:91]: nit: i will recommend to put `clientPool` as the last arg here

File: table/server/underdb/hive/src/main/java/alluxio/table/under/hive/util/HiveClientPool.java [hunk 3]
```diff
@@ -11,105 +11,23 @@
...
-    // there is no way to check if a hive client is connected.
-    // TODO(gpang): periodically and asynchronously check the health of clients
-    return true;
-  }
-
-  @Override
-  protected boolean shouldGc(ResourceInternal<IMetaStoreClient> clientResourceInternal) {
-    return System.currentTimeMillis() - clientResourceInternal
-        .getLastAccessTimeMs() > mGcThresholdMs;
+public abstract class HiveClientPool extends DynamicResourcePool<IMetaStoreClient> {
```
Comment (apc999) [inline table/server/underdb/hive/src/main/java/alluxio/table/under/hive/util/HiveClientPool.java:24]: nit: `AbstractHiveClientPool`

**Review by apc999 — CHANGES_REQUESTED** [review]
Minor comments left. Overall looks good to me

**Comment by dbw9580** [comment]
Thanks for the comments @apc999. Revised in 15c42da342a927c9b79b8443abfbd2465d083cbd.

**Review by HelloHorizon — APPROVED** [review]
@dbw9580 Thanks for the improvement! LGTM

**Comment by HelloHorizon** [comment]
alluxio-bot, merge this please

**Pull request merged**